cub/cub/thread/thread_reduce.cuh

```diff
@@ -105,101 +206,115 @@ ThreadReduce(T* input, ReductionOp reduction_op, PrefixT prefix, Int2Type<LENGTH
- *   Input array
+ *   Input pointer
  *
  * @param[in] reduction_op
  *   Binary reduction operator
+ *
+ * @return Aggregate of type <tt>cuda::std::__accumulator_t<ReductionOp, T></tt>
  */
-template <int LENGTH, typename T, typename ReductionOp>
-_CCCL_DEVICE _CCCL_FORCEINLINE T ThreadReduce(T* input, ReductionOp reduction_op)
+template <int LENGTH, typename T, typename ReductionOp, typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, T>>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE AccumT ThreadReduce(const T* input, ReductionOp reduction_op)
 {
-  T prefix = input[0];
-  return ThreadReduce<LENGTH - 1>(input + 1, reduction_op, prefix);
+  static_assert(LENGTH > 0, "LENGTH must be greater than 0");
+  static_assert(detail::has_binary_operator<ReductionOp, T>::value,
+                "ReductionOp must have the binary operator: operator(V1, V2)");
+  using ArrayT = T[LENGTH];
+  auto& array  = reinterpret_cast<const ArrayT&>(input);
+  return ThreadReduce(array, reduction_op);
 }
 
 /**
- * @brief Perform a sequential reduction over the statically-sized @p input array,
- *        seeded with the specified @p prefix. The aggregate is returned.
+ * @brief Perform a sequential reduction over @p LENGTH elements of the @p input pointer, seeded with the specified @p
+ *        prefix. The aggregate is returned.
  *
  * @tparam LENGTH
- *   <b>[inferred]</b> LengthT of @p input array
+ *   Length of input pointer
  *
  * @tparam T
- *   <b>[inferred]</b> The data type to be reduced.
+ *   <b>[inferred]</b> The data type to be reduced
  *
  * @tparam ReductionOp
  *   <b>[inferred]</b> Binary reduction operator type having member
  *   <tt>T operator()(const T &a, const T &b)</tt>
  *
+ * @tparam PrefixT
+ *   <b>[inferred]</b> The prefix type
+ *
  * @param[in] input
- *   Input array
+ *   Input pointer
  *
  * @param[in] reduction_op
  *   Binary reduction operator
  *
  * @param[in] prefix
  *   Prefix to seed reduction with
+ *
+ * @return Aggregate of type <tt>cuda::std::__accumulator_t<ReductionOp, T, PrefixT></tt>
  */
 template <int LENGTH,
           typename T,
           typename ReductionOp,
           typename PrefixT,
-          typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, T, PrefixT>>
-_CCCL_DEVICE _CCCL_FORCEINLINE AccumT ThreadReduce(T (&input)[LENGTH], ReductionOp reduction_op, PrefixT prefix)
+          typename AccumT = ::cuda::std::__accumulator_t<ReductionOp, T, PrefixT>,
+          _CUB_TEMPLATE_REQUIRES(LENGTH > 0)>
+_CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE AccumT
+ThreadReduce(const T* input, ReductionOp reduction_op, PrefixT prefix)
 {
-  return ThreadReduce(input, reduction_op, prefix, Int2Type<LENGTH>());
+  static_assert(detail::has_binary_operator<ReductionOp, T>::value,
+                "ReductionOp must have the binary operator: operator(V1, V2)");
+  using ArrayT = T[LENGTH];
+  auto& array  = reinterpret_cast<const ArrayT&>(input);
+  return ThreadReduce(array, reduction_op, prefix);
 }
```

Comment:
if I'm understand correctly, you are asking why we have an interface with a pointer instead of an array-like object. There are several cases in CUB where the code uses this interface.

Reply:
> if I'm understand correctly, you are asking why we have an interface with a pointer instead of an array-like object.

Yes. The API used to take a `T (&input)[LENGTH]`, but now it takes a `const T* input`. Now, the template parameter `LENGTH` can no l longer be deduced and needs to be specified at every callsite, which is annoying.

Reply:
that's a good suggestion. I need to check if possible because there are cases with LENGTH == 0 and array of size 0 is a gcc extension

Reply:
I would keep removing the pointer interface for another PR because it requires to change the code in multiple places. I will add the deprecated remark

Reply:
> I will add the deprecated remark

No need, it's an internal API! Just make sure you refactor all callsites!